### ISSUE TITLE
update: when using --to-remote, only update missing cache files

### ISSUE THREAD

**Issue by isidentical** [iterative/dvc#5768]
Currently what `update --to-remote` does is the same exact thing with `import-url --to-remote` with only except that update is using the stage's url instead of taking it as an argument. This means that for each source file, even though they might already exist on our remote storage: we download them, calculate their hashes, and upload them to the remote storage. 

For `import --to-remote`; downloading, uploading, and hash calculation happens at the same time, which means we only do `1 download + 1 upload `. Though this is not as efficient as it might get for `update --to-remote` with a directory that consists of small files. For each file, do the same actions without ever checking whether they are present or not. Since we can not determine their presence beforehand (before actually downloading the file and calculating hash in chunks), we can optimize the `update` in a way that works for cases that the updated files are always less than total amount of files for a directory that consist of small files.

So instead of doing `1 download + 1 upload` on every case,  we can do `1 download` on best cases (when the file already exists) and `2 download + 1 upload` for worst cases (when the file doesn't exist on the cache).  This might be a bad optimization when the changes on the source is always bigger than the remaining, unchanged part though I think it is a bit unusual and rare case.

CC: @efiop @dberenbaum

**Comment by dberenbaum** [comment]
> So instead of doing `1 download + 1 upload` on every case, we can do `1 download` on best cases (when the file already exists) and `2 download + 1 upload` for worst cases (when the file doesn't exist on the cache).

Sorry, @isidentical, can you explain in a bit more detail what's happening in each case? I'm not following why an extra download is required if the file doesn't exist.

**Comment by isidentical** [comment]
> I'm not following why an extra download is required if the file doesn't exist.

For finding the only missing files, we need to calculate the hashes and find out which one of those hashes are not present in the remote. Unfortunately, we can not calculate the hashes without downloading all the files beforehand. So instead of wasting an upload on a file that might exist, we initially download everything (actually we don't download anything to anywhere, just calculate the hashes, but we waste the same time) and then find out the missing hashes and apply the normal procedure (like this time download normally as in chunks and upload that chunks to the remote). 

```
import-url --to-remote => download [from url] -> md5 -> upload [to remote storage]
update => download [from url] -> md5 -> download [from url, but only missing files on the remote storage] -> upload [to remote storage]
```

```

**Comment by dberenbaum** [comment]
How does the md5 step work? Is it calculated per file? Is each file read into memory in its entirety?

**Comment by dberenbaum** [comment]
Also, how does this relate to #4889? Is it the `--to-remote` equivalent of that ticket?

**Comment by isidentical** [comment]
> Also, how does this relate to #4889? Is it the --to-remote equivalent of that ticket?

No, not really

**Comment by isidentical** [comment]
> How does the md5 step work? Is it calculated per file? Is each file read into memory in its entirety?

It can be thought as mapping the entire remote url first, and then by comparing that map with the remote storage finding the missing ones. It is calculated per file basis, and only 1 chunk of a file (platform-dependant chunk size) stored on the memory at the time (* number of jobs, since we are working concurrently)

**Comment by dberenbaum** [comment]
> > Also, how does this relate to #4889? Is it the --to-remote equivalent of that ticket?
> 
> No, not really

Could you elaborate on whether they are related and how? I don't know the details well enough, so it seems to me like:

* This ticket is about optimizing `update --to-remote` so that files already on remote are not uploaded again.
* #4889 is about optimizing `update` so that files already in cache are not downloaded again.

**Comment by dberenbaum** [comment]
If only missing files are updated on the remote, how would dvc check for files already on the remote that were deleted/changed in the url?

**Comment by isidentical** [comment]
> This ticket is about optimizing update --to-remote so that files already on remote are not uploaded again.

The internals of to-remote is much different than the bare import-url/update, so this ticket's approach is specific to some presumptions that are ensured by the underlying design of `--to-remote`. You probably don't need much of the extra stuff we are doing here for regular import-url/update (since you are not bound to md5s / not presented data), so I think they are different (unless I am missing something, I don't know much about that ticket but from what I see).

> If only missing files are updated on the remote, how would dvc check for files already on the remote that were deleted/changed in the url?

So basically what we do is we first compute all the necessary hashes for the url, like `{'a/b/c': 'some_md5', 'c/d/e': 'some_other_md5'}` (this stage costs a single download for every file on the remote). After this mapping, we check the remote for the existence of those md5s and eliminate the existing hashes (so if you have 'some_md5' on the remote storage, we won't try to update 'a/b/c' again) and transfer the rest (spending 1 download + 1 upload for the transfer). So if the large part of hashes present on the remote, then this would save you `N-number_of_new_cache_files` (900) files the amount of uploads (e.g out of 1000, 900 is already there 100 is new) and make you waste `number_of_new_cache_files` (100) on downloads.

**Comment by isidentical** [comment]
With #5773, the performance of `update` is substantially improved;
|                            | before #5773 | after #5773 |
|----------------------------|--------------|-------------|
| dvc import-url --to-remote | 3m48,789s    | 2m10,561s   |
| dvc update --to-remote     | 9m16,278s    | 2m49,857s   |

A few future steps:
- [ ] Benefit from `verify` option
- [ ] Reduce the code duplication for different parts (add --to-remote/import-url --to-remote/update --to-remote) with implementing a common `out.transfer()` method

**Comment by isidentical** [comment]
|                            | before #5860 | after #5860 |
|----------------------------|--------------|-------------|
| dvc import-url --to-remote | unchanged    | unchanged   |
| dvc update --to-remote     | 2m49,857s    | 2m24,000s   |

(update syncs %50 more data, and with that I think we are in a very good shape)

**Comment by efiop** [comment]
> (update syncs %50 more data, and with that I think we are in a very good shape)

For the record: there are a few more tricks up our sleeves, e.g. not only using `verify` but also leveraging indexes that we have. So it will be cut even more. ;)

**Comment by isidentical** [comment]
> also leveraging indexes that we have

What do you mean by indexes? state?

**Comment by efiop** [comment]
@isidentical I mean `dvc/remote/index.py`. E.g. for old `12345.dir` hash, if that object exists on s3, we can assume that every file in that directory already exists on the remote too. That's what we use for pull/push/status right now.

**Comment by isidentical** [comment]
> I mean dvc/remote/index.py. E.g. for old 12345.dir hash, if that object exists on s3, we can assume that every file in that directory already exists on the remote too. That's what we use for pull/push/status right now.

Ah, I actually had a patch (in the old transfer() implementation) for that (with `known_hash`, maybe you remember). I might check later about whether I can port it to the current version or not.

**Comment by efiop** [comment]
@isidentical Oh, that's great! It will come in handy in the next steps for odb, let's discuss them during our next meeting. We'll need to port old dvc/remote logic to the new odb arch anyway, and all of those optimizations should automatically kick in for both push/pull/status/etc and --to-remote/to-cache.

**Comment by efiop** [comment]
@isidentical Though please feel free to send a WIP if you feel like it, it will help the discussion, maybe we are on the same page already ;)

**Comment by isidentical** [comment]
Closing this issue since it's initial goals (and some extras) are done, and porting the `dvc/remote` logic to the new odb arch definitely will have its own separate issue.
